### ISSUE TITLE
Infinite loop with insert / emplace on MSVC

### ISSUE THREAD

**Issue by jcelerier** [Tessil/hopscotch-map#33]
Certainly the fault of a dumb compiler, but still : given `hopscotch_map<void*>`, I get an infinite recursion between these two functions: 

    template<class P, typename std::enable_if<std::is_constructible<value_type, P&&>::value>::type* = nullptr>
    std::pair<iterator, bool> insert(P&& value) { 
        return emplace(std::forward<P>(value)); 
    }
    
    template<class... Args>
    std::pair<iterator, bool> emplace(Args&&... args) {
        return insert(value_type(std::forward<Args>(args)...));
    }

**Comment by Tessil** [comment]
Thank you for the report.

I'm confused that emplace doesn't call `std::pair<iterator, bool> insert(value_type&& value)`, it should have priority over `template<class P, typename std::enable_if<std::is_constructible<value_type, P&&>::value>::type* = nullptr> std::pair<iterator, bool> insert(P&& value)` as emplace calls insert with a rvalue `value_type`. Probably a bug in the compiler effectively.

I'll try to test that when I have more time (it probably also affects my other projects) and merge your pull request.

**Comment by jcelerier** [comment]
yes, this did not really make sense to me either, and it works fine with clang & gcc.

**Comment by Tessil** [comment]
Do you have a small example?

I tried to reproduce the bug with the following code on VS2015 (Update 3):
```c++
int a = 10;

tsl::hopscotch_map<void*, void*> map;
std::pair<int*, int*> p = {&a, &a};
map.insert(p);
```

But it works as intended.

**Comment by jcelerier** [comment]
Here's a repro (using VS 2017 15.7.3)

    #include <vector>
    #include <hopscotch_set.h>
    int main()
    {
      std::vector<int*> lst;
      tsl::hopscotch_set<int*> inlets;
      inlets.insert(lst.begin(), lst.end());
    }

VS is kind enough to show it as a a warning : 

    1>c:\users\blah\score\api\3rdparty\hopscotch-map\src\hopscotch_hash.h(1119): warning C4717: 'tsl::detail_hopscotch_hash::hopscotch_hash<int *,tsl::hopscotch_set<int *,std::hash<int *>,std::equal_to<int *>,std::allocator<int *>,62,0,tsl::power_of_two_growth_policy>::KeySelect,void,std::hash<int *>,std::equal_to<int *>,std::allocator<int *>,62,0,tsl::power_of_two_growth_policy,std::list<int *,std::allocator<int *> > >::emplace<int * &>': recursive on all control paths, function will cause runtime stack overflow
    1>c:\users\blah\score\api\3rdparty\hopscotch-map\src\hopscotch_hash.h(1029): warning C4717: 'tsl::detail_hopscotch_hash::hopscotch_hash<int *,tsl::hopscotch_set<int *,std::hash<int *>,std::equal_to<int *>,std::allocator<int *>,62,0,tsl::power_of_two_growth_policy>::KeySelect,void,std::hash<int *>,std::equal_to<int *>,std::allocator<int *>,62,0,tsl::power_of_two_growth_policy,std::list<int *,std::allocator<int *> > >::insert<int * &,0>': recursive on all control paths, function will cause runtime stack overflow

Compiling with /O2 is enough to reproduce.

**Comment by Tessil** [comment]
Thank you, I was able to reproduce the bug. Though I only have a warning on my side with MSVC 2015, update 3. When using emplace the code itself runs fine without recursion.

I merged your PR.

**Comment by jcelerier** [comment]
thanks !
